src/tncdr/stabilizer_mps/tensor_network.py

```diff
@@ -98,21 +103,120 @@ def contract(self, node_in:str, node_out:str, edge_ids:Union[str, list[str]], ne
             self.remove_edge(node_in=node_in, node_out=node_out, edge_id=edge_id)
 
         if node_in==node_out:
-            return self._partial_trace(
+            self._partial_trace(
                 node=node_in,
-                new_node_id=new_node_id,
+                new_node_id=0,
                 directions_in=directions_in,
                 directions_out=directions_out
             )
         else:
-            return self._contract_separate_nodes(
+            self._contract_separate_nodes(
                 node_in=node_in,
                 node_out=node_out,
-                new_node_id=new_node_id,
+                new_node_id=0,
                 directions_in=directions_in,
                 directions_out=directions_out
             )
 
+        nx.relabel_nodes(self.tensornet, {0:new_node_id}, copy=False)
+
+    def svd_decomposition(
+            self, 
+            node:str,
+            left_node_id:str, 
+            left_node_edges:Union[str, list[str]],
+            right_node_id:str, 
+            right_node_edges:Union[str, list[str]],
+            middle_node_id:str,
+            middle_edge:str,
+            bond_dimension:Optional[int]=None,
+        ):
+        
+        if type(left_node_edges) is str: 
+            left_node_edges = [left_node_edges]
+        
+        if type(right_node_edges) is str: 
+            right_node_edges = [right_node_edges]
+        
+        tensor = self.tensornet.nodes[node]['tensor']
+        
+        # Transpose and reshape into a matrix
+        transposition_vector = self._svd_transposition_vector(node, left_node_edges, right_node_edges)
+        np.transpose(tensor, transposition_vector)
```

Comment:
We probably have to pay attention to this step. If your idea was to transpose the original vector through `np.transpose`, I am afraid it is not happening here: it seems the `np.transpose` function is returning an unchanged view of the tensor (not modifying it in-place). Thus, I guess in the next lines we are operating on the old - non transposed - version of `tensor` [here the link to Numpy docs](https://numpy.org/doc/1.26/reference/generated/numpy.transpose.html).

```suggestion
        tensor = np.ascontiguousarray(np.transpose(tensor, transposition_vector))
```

As an extra suggestion we can save it as [contiguous array](https://numpy.org/doc/1.26/reference/generated/numpy.ascontiguousarray.html#numpy-ascontiguousarray) in memory (don't know very well why, but it seems this can be useful when performing this kind of algebra).

Reply:
You are right, I forgot the assignment to tensor. Good catch!